base/abstractarray.jl

```diff
@@ -126,6 +126,34 @@ _length(A) = length(A)
 endof(a::AbstractArray) = length(a)
 first(a::AbstractArray) = a[first(eachindex(a))]
 
+function fixate_eltype{T<:AbstractArray}(::Type{T}, default::Type)
+    @_pure_meta
+    aa_type = T
+    while aa_type.name.primary != AbstractArray
+        aa_type = supertype(aa_type)
+    end
+    if isa(aa_type.parameters[1], TypeVar)
+        varname = aa_type.parameters[1].name
+        new_params = collect(T.parameters)
+        for i in eachindex(new_params)
+            p = new_params[i]
+            if isa(p, TypeVar) && p.name == varname
+                new_params[i] = default
+            end
+        end
+        return T.name.primary{new_params...}
+    else
+        return T
+    end
+end
```

Comment:
I see. I had in mind another use case, using something like this combined with something like `deepcopy_internal` to write a generic `similar(::AbstractArray, ::Type)`. Handling a dimensions in a generic fashion might not be possible and we still would need specialization (as you have here).

With this in mind and what you is needed a third boolean argument would seem like the way to cover both scenarios.